### PR TITLE
Add HPA for web server to manage Load

### PR DESCRIPTION
**Pull request by ranjith-ka**
Issue Endorsed by Maintainers
Description of the Change:

- Add HPA for web server. 
- Remove Namespace , since Helm3 use namespace outside the templates. 

Alternate Designs

Possible Drawbacks

Verification Process

Release Notes

### CODE REVIEW

**Comment by marcofranssen** [comment]
Thanks for your contribution @ranjith-ka.

Could you please run the following script and commit those changes to have CI pass.

```bash
.github/helm-docs.sh
```

File: templates/web-serviceaccount.yaml [hunk 1]
```diff
@@ -3,7 +3,6 @@ apiVersion: v1
 kind: ServiceAccount
 metadata:
   name: {{ include "dctna.serviceAccountName" . }}-web
-  namespace: {{ .Release.Namespace }}
```
Comment (marcofranssen) [inline templates/web-serviceaccount.yaml:6]: please revert this change.

File: templates/web-service.yaml [hunk 2]
```diff
@@ -2,7 +2,6 @@ apiVersion: v1
 kind: Service
 metadata:
   name: {{ include "dctna.fullname" . }}-web
-  namespace: {{ .Release.Namespace }}
```
Comment (marcofranssen) [inline templates/web-service.yaml:5]: please revert this change.

File: templates/web-deployment.yaml [hunk 3]
```diff
@@ -2,11 +2,12 @@ apiVersion: apps/v1
 kind: Deployment
 metadata:
   name: {{ include "dctna.fullname" . }}-web
-  namespace: {{ .Release.Namespace }}
```
Comment (marcofranssen) [inline templates/web-deployment.yaml:5]: please revert this change.

File: templates/server-statefulset.yaml [hunk 4]
```diff
@@ -2,7 +2,6 @@ apiVersion: apps/v1
 kind: StatefulSet
 metadata:
   name: {{ include "dctna.fullname" . }}-server
-  namespace: {{ .Release.Namespace }}
```
Comment (marcofranssen) [inline templates/server-statefulset.yaml:5]: please revert this change.

File: templates/server-serviceaccount.yaml [hunk 5]
```diff
@@ -3,7 +3,6 @@ apiVersion: v1
 kind: ServiceAccount
 metadata:
   name: {{ include "dctna.serviceAccountName" . }}-server
-  namespace: {{ .Release.Namespace }}
```
Comment (marcofranssen) [inline templates/server-serviceaccount.yaml:6]: please revert this change.

File: templates/server-service.yaml [hunk 6]
```diff
@@ -2,7 +2,6 @@ apiVersion: v1
 kind: Service
 metadata:
   name: {{ include "dctna.fullname" . }}-server
-  namespace: {{ .Release.Namespace }}
```
Comment (marcofranssen) [inline templates/server-service.yaml:5]: please revert this change.

File: templates/server-notary-config-configmap.yml [hunk 7]
```diff
@@ -2,7 +2,6 @@ apiVersion: v1
 kind: ConfigMap
 metadata:
   name: {{ include "dctna.fullname" . }}-notary-config
-  namespace: {{ .Release.Namespace }}
```
Comment (marcofranssen) [inline templates/server-notary-config-configmap.yml:5]: please revert this change.

File: templates/notary-auth-secret.yml [hunk 8]
```diff
@@ -3,7 +3,6 @@ apiVersion: v1
 kind: Secret
 metadata:
   name: {{ include "dctna.fullname" . }}-notary-auth
-  namespace: {{ .Release.Namespace }}
```
Comment (marcofranssen) [inline templates/notary-auth-secret.yml:6]: please revert this change.

File: README.md [hunk 9]
```diff
@@ -77,3 +77,10 @@ Kubernetes: `>=1.15.0-0`
 | web.service.port | int | `80` |  |
 | web.service.type | string | `"ClusterIP"` |  |
 | web.tolerations | list | `[]` |  |
+| web.livenessProbe.enabled | bool | `true` | |
```
Comment (marcofranssen) [inline README.md:80]: This can be generated using the `.github/helm-docs.sh` script. This is also the reason for CI to fail on this repo.

File: templates/web-deployment.yaml [hunk 10]
```diff
@@ -41,14 +42,24 @@ spec:
...
+            timeoutSeconds: {{ .Values.web.livenessProbe.timeoutSeconds | default 20 }}
+          {{- end }}
+          {{- if .Values.web.readinessProbe.enabled }}
           readinessProbe:
             httpGet:
-              path: /
-              port: http
+              path: {{ .Values.web.readinessProbe.path | default "/" }}
+              port: {{ .Values.web.service.port | default "http" }}
+            initialDelaySeconds: {{ .Values.web.readinessProbe.initialDelaySeconds | default 10 }}
```
Comment (marcofranssen) [inline templates/web-deployment.yaml:58]:
```suggestion
            initialDelaySeconds: {{ .Values.web.readinessProbe.initialDelaySeconds | default 5 }}
```

This container is a static web app running in nginx which boots within few seconds so the delay can be shorter to have a quicker response and autoscaling.

File: templates/web-deployment.yaml [hunk 11]
```diff
@@ -41,14 +42,24 @@ spec:
...
+            timeoutSeconds: {{ .Values.web.readinessProbe.timeoutSeconds | default 10 }}
```
Comment (marcofranssen) [inline templates/web-deployment.yaml:59]:
```suggestion
            timeoutSeconds: {{ .Values.web.readinessProbe.timeoutSeconds | default 5 }}
```

File: values.yaml [hunk 12]
```diff
@@ -16,6 +16,16 @@ serviceAccount:
...
 web:
+  livenessProbe:
+    enabled: true
+  readinessProbe:
+    enabled: true
+  autoscaling:
+    enabled: true
+    minReplicas: 1
+    maxReplicas: 5
+    targetCPUUtilizationPercentage: 60
```
Comment (marcofranssen) [inline values.yaml:27]:
```suggestion
    targetCPUUtilizationPercentage: 80
```

File: values.yaml [hunk 13]
```diff
@@ -16,6 +16,16 @@ serviceAccount:
...
+    targetMemoryUtilizationPercentage: 60
```
Comment (marcofranssen) [inline values.yaml:28]:
```suggestion
    targetMemoryUtilizationPercentage: 80
```

As this container boots really quick and with the adjusted timeouts as suggested above we can slightly delay the autoscaling to happen only at 80% as default values for this chart.

**Review by marcofranssen — CHANGES_REQUESTED** [review]
Please see the review. for some suggested changes.

File: templates/server-serviceaccount.yaml (hunk 5 above)
Comment (ranjith-ka) [reply templates/server-serviceaccount.yaml:6]: But Helm3 we manage the NameSpace outside the helm templates?  I am missing something ?

Comment (marcofranssen) [reply templates/server-serviceaccount.yaml:6]: when using `helm template` the namespaces will be in the kubectl files allowing them to be used without defining a namespace. The namespace can still be overridden using the `-n namespace`. The dicission to keep or remote these namespaces I would like to keep separate from this PR.

**Review by marcofranssen — APPROVED** [review]

**Review by marcofranssen — APPROVED** [review]
Will merge soon once I have the time to test on our cluster. Thanks a lot for your contribution.
